[SUPPORT] Scala 2.12 UnitTest with HUDI 0.15.0 / CloudWatchMetricsReporter issue

**Describe the problem you faced**

When I try to run a UnitTest on my Scala project using HUDI and the CloudWatch metrics option to "true" I get the following exception,

`org.apache.hudi.exception.HoodieException: Unable to instantiate class org.apache.hudi.metrics.cloudwatch.CloudWatchMetricsReporter`


**To Reproduce**

Steps to reproduce the behavior:

1. Use this "dummy" UnitTest
```
package xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx

import org.apache.hudi.DataSourceWriteOptions
import org.apache.hudi.config.HoodieWriteConfig
import org.apache.spark.sql.types._
import org.apache.spark.sql.{DataFrame, SparkSession}
import org.junit.runner.RunWith
import org.scalatest.BeforeAndAfterAll
import org.scalatest.funsuite.AnyFunSuite
import org.scalatestplus.junit.JUnitRunner


@RunWith(classOf[JUnitRunner])
class HudiWriteReadTest extends AnyFunSuite with BeforeAndAfterAll {

  // Initialize Spark session
  val spark: SparkSession = SparkSession
    .builder()
    .master("local[1]")
    .appName("xxxxxxxxxxxxxxxxx.HudiWriteReadTest")
    .config("spark.serializer", "org.apache.spark.serializer.KryoSerializer")
    .config("spark.sql.catalog.spark_catalog", "org.apache.spark.sql.hudi.catalog.HoodieCatalog")
    .config("spark.sql.extensions", "org.apache.spark.sql.hudi.HoodieSparkSessionExtension")
    .getOrCreate()

  import spark.implicits._

  // Define schema
  val schema = StructType(
    Seq(
      StructField("id", IntegerType, false),
      StructField("name", StringType, false),
      StructField("timestamp", LongType, false)
    )
  )

  def areDataFramesEqual(df1: DataFrame, df2: DataFrame): Boolean = {
    // Define the columns to compare (excluding metadata columns)
    val columnsToCompare = Seq("id", "name", "timestamp")

    // Select only the columns to compare
    val df1Selected = df1.select(columnsToCompare.head, columnsToCompare.tail: _*)
    val df2Selected = df2.select(columnsToCompare.head, columnsToCompare.tail: _*)

    // Check if schemas are identical
    if (!df1Selected.schema.equals(df2Selected.schema)) {
      println("Schemas do not match:")
      println("DF1 Schema: " + df1Selected.schema.treeString)
      println("DF2 Schema: " + df2Selected.schema.treeString)
      return false
    }

    // Collect rows from both DataFrames
    val rows1 = df1Selected.collect()
    val rows2 = df2Selected.collect()

    // Check if the number of rows is the same
    if (rows1.length != rows2.length) {
      println(s"Row counts differ: DF1 has ${rows1.length} rows, DF2 has ${rows2.length} rows.")
      return false
    }

    // Sort rows by a unique identifier (e.g., "id") to ensure consistent ordering
    val sortedRows1 = rows1.sortBy(_.getAs[Int]("id"))
    val sortedRows2 = rows2.sortBy(_.getAs[Int]("id"))

    // Compare sorted rows
    sortedRows1.zip(sortedRows2).forall {
      case (row1, row2) => row1 == row2
    }
  }

  // Function to write DataFrame to Hudi
  def writeToHudi(df: DataFrame, tablePath: String): Unit = {
    df.write
      .format("hudi")
      .option(DataSourceWriteOptions.TABLE_TYPE_OPT_KEY, "COPY_ON_WRITE")
      .option(DataSourceWriteOptions.OPERATION_OPT_KEY, "insert")
      .option(DataSourceWriteOptions.RECORDKEY_FIELD_OPT_KEY, "id")
      .option(DataSourceWriteOptions.PRECOMBINE_FIELD_OPT_KEY, "timestamp")
      .option(DataSourceWriteOptions.PARTITIONPATH_FIELD_OPT_KEY, "")
      .option(HoodieWriteConfig.TABLE_NAME, "hudi_test_table")
      .option("hoodie.metrics.on", "true")
      .option("hoodie.metrics.reporter.type", "CLOUDWATCH")
      .mode("overwrite")
      .save(tablePath)
  }

  // Function to read DataFrame from Hudi
  def readFromHudi(tablePath: String): DataFrame = {
    spark.read
      .format("hudi")
      .load(tablePath)
  }

  // Unit test
  test("Hudi write and read operations") {
    val data = Seq(
      (1, "Alice", 1000L),
      (2, "Bob", 2000L)
    )
    val df        = data.toDF("id", "name", "timestamp")
    val tablePath = "file:///tmp/hudi_test_table"

    // Write data to Hudi
    writeToHudi(df, tablePath)

    // Read data from Hudi
    val readDf = readFromHudi(tablePath)

    // Validate the data
    assert(areDataFramesEqual(readDf, df))
  }

  // Stop Spark session after tests
  override def afterAll(): Unit = {
    spark.stop()
  }
}

```

2. Run the ./gradlew build
3. Get the Exception
`org.apache.hudi.exception.HoodieException: Unable to instantiate class org.apache.hudi.metrics.cloudwatch.CloudWatchMetricsReporter
`

**Expected behavior**

The UnitTest should run fine. If I set the CloudWatch to "false", it works.

**Environment Description**

* Hudi version : 0.15.0

* Spark version : 3.5.2

* Hive version : spark-hive_2.12:3.5.2

* Hadoop version : 3.4.0

* Storage (HDFS/S3/GCS..) :   Local on PC

* Running on Docker? (yes/no) : no


**Additional context**

This is the build.gradle I'm using

dependencies {
    // Implementation dependencies
    implementation "org.apache.hadoop:hadoop-aws:3.4.0"
    implementation "org.apache.hadoop:hadoop-common:3.4.0"
    implementation "org.apache.spark:spark-sql-kafka-0-10_2.12:3.5.2"
    implementation "org.apache.spark:spark-avro_2.12:3.5.2"
    implementation "org.apache.hudi:hudi-spark3.5-bundle_2.12:0.15.0"
    implementation 'org.apache.hudi:hudi-aws-bundle:0.15.0'
    implementation "com.amazonaws:aws-java-sdk-s3:1.12.772"
    implementation "com.amazonaws:aws-java-sdk-secretsmanager:1.12.772"
    implementation 'io.logz.sawmill:sawmill-core:2.0.20'
    implementation "org.apache.spark:spark-core_2.12:3.5.2"
    implementation "com.amazonaws:aws-java-sdk-cloudwatch:1.12.772"

    // Compile-only dependencies
    compileOnly "org.scala-lang:scala-library:2.12.8"
    compileOnly "com.amazonaws:aws-java-sdk-secretsmanager:1.12.335"
    compileOnly "org.apache.spark:spark-hive_2.12:3.5.2"

    // Test dependencies
    testImplementation "org.scalatest:scalatest_2.12:3.2.19"
    testImplementation "junit:junit:4.13.2"
    testImplementation "org.scalatestplus:junit-4-13_2.12:3.2.15.0"
}


**Stacktrace**

```
org.apache.hudi.exception.HoodieException: Unable to instantiate class org.apache.hudi.metrics.cloudwatch.CloudWatchMetricsReporter
	at org.apache.hudi.common.util.ReflectionUtils.loadClass(ReflectionUtils.java:75)
	at org.apache.hudi.metrics.MetricsReporterFactory.createReporter(MetricsReporterFactory.java:89)
	at org.apache.hudi.metrics.Metrics.<init>(Metrics.java:61)
	at org.apache.hudi.metrics.Metrics.getInstance(Metrics.java:87)
	at org.apache.hudi.metrics.HoodieMetrics.<init>(HoodieMetrics.java:101)
	at org.apache.hudi.client.BaseHoodieClient.<init>(BaseHoodieClient.java:101)
	at org.apache.hudi.client.BaseHoodieWriteClient.<init>(BaseHoodieWriteClient.java:164)
	at org.apache.hudi.client.SparkRDDWriteClient.<init>(SparkRDDWriteClient.java:90)
	at org.apache.hudi.client.SparkRDDWriteClient.<init>(SparkRDDWriteClient.java:74)
	at org.apache.hudi.DataSourceUtils.createHoodieClient(DataSourceUtils.java:208)
	at org.apache.hudi.HoodieSparkSqlWriterInternal.$anonfun$writeInternal$27(HoodieSparkSqlWriter.scala:451)
	at scala.Option.getOrElse(Option.scala:189)
	at org.apache.hudi.HoodieSparkSqlWriterInternal.writeInternal(HoodieSparkSqlWriter.scala:448)
	at org.apache.hudi.HoodieSparkSqlWriterInternal.write(HoodieSparkSqlWriter.scala:187)
	at org.apache.hudi.HoodieSparkSqlWriter$.write(HoodieSparkSqlWriter.scala:125)
	at org.apache.hudi.DefaultSource.createRelation(DefaultSource.scala:168)
	at org.apache.spark.sql.execution.datasources.SaveIntoDataSourceCommand.run(SaveIntoDataSourceCommand.scala:48)
	at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult$lzycompute(commands.scala:75)
	at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult(commands.scala:73)
	at org.apache.spark.sql.execution.command.ExecutedCommandExec.executeCollect(commands.scala:84)
	at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$1.$anonfun$applyOrElse$1(QueryExecution.scala:107)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$6(SQLExecution.scala:125)
	at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:201)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:108)
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:900)
	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:66)
	at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$1.applyOrElse(QueryExecution.scala:107)
	at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$1.applyOrElse(QueryExecution.scala:98)
	at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$transformDownWithPruning$1(TreeNode.scala:461)
	at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(origin.scala:76)
	at org.apache.spark.sql.catalyst.trees.TreeNode.transformDownWithPruning(TreeNode.scala:461)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.org$apache$spark$sql$catalyst$plans$logical$AnalysisHelper$$super$transformDownWithPruning(LogicalPlan.scala:32)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.transformDownWithPruning(AnalysisHelper.scala:267)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.transformDownWithPruning$(AnalysisHelper.scala:263)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.transformDownWithPruning(LogicalPlan.scala:32)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.transformDownWithPruning(LogicalPlan.scala:32)
	at org.apache.spark.sql.catalyst.trees.TreeNode.transformDown(TreeNode.scala:437)
	at org.apache.spark.sql.execution.QueryExecution.eagerlyExecuteCommands(QueryExecution.scala:98)
	at org.apache.spark.sql.execution.QueryExecution.commandExecuted$lzycompute(QueryExecution.scala:85)
	at org.apache.spark.sql.execution.QueryExecution.commandExecuted(QueryExecution.scala:83)
	at org.apache.spark.sql.execution.QueryExecution.assertCommandExecuted(QueryExecution.scala:142)
	at org.apache.spark.sql.DataFrameWriter.runCommand(DataFrameWriter.scala:859)
	at org.apache.spark.sql.DataFrameWriter.saveToV1Source(DataFrameWriter.scala:388)
	at org.apache.spark.sql.DataFrameWriter.saveInternal(DataFrameWriter.scala:361)
	at org.apache.spark.sql.DataFrameWriter.save(DataFrameWriter.scala:240)
	at com.anonymized.cde.smmdevicetype.HudiWriteReadTest.writeToHudi(HudiWriteReadTest.scala:87)
	at com.anonymized.cde.smmdevicetype.HudiWriteReadTest.$anonfun$new$1(HudiWriteReadTest.scala:107)
	at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
	at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
	at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
	at org.scalatest.Transformer.apply(Transformer.scala:22)
	at org.scalatest.Transformer.apply(Transformer.scala:20)
	at org.scalatest.funsuite.AnyFunSuiteLike$$anon$1.apply(AnyFunSuiteLike.scala:226)
	at org.scalatest.TestSuite.withFixture(TestSuite.scala:196)
	at org.scalatest.TestSuite.withFixture$(TestSuite.scala:195)
	at org.scalatest.funsuite.AnyFunSuite.withFixture(AnyFunSuite.scala:1564)
	at org.scalatest.funsuite.AnyFunSuiteLike.invokeWithFixture$1(AnyFunSuiteLike.scala:224)
	at org.scalatest.funsuite.AnyFunSuiteLike.$anonfun$runTest$1(AnyFunSuiteLike.scala:236)
	at org.scalatest.SuperEngine.runTestImpl(Engine.scala:306)
	at org.scalatest.funsuite.AnyFunSuiteLike.runTest(AnyFunSuiteLike.scala:236)
	at org.scalatest.funsuite.AnyFunSuiteLike.runTest$(AnyFunSuiteLike.scala:218)
	at org.scalatest.funsuite.AnyFunSuite.runTest(AnyFunSuite.scala:1564)
	at org.scalatest.Suite.run(Suite.scala:1114)
	at org.scalatest.Suite.run$(Suite.scala:1096)
	at org.scalatest.funsuite.AnyFunSuite.org$scalatest$funsuite$AnyFunSuiteLike$$super$run(AnyFunSuite.scala:1564)
	at org.scalatest.funsuite.AnyFunSuiteLike.$anonfun$run$1(AnyFunSuiteLike.scala:273)
	at org.scalatest.SuperEngine.runImpl(Engine.scala:535)
	at org.scalatest.funsuite.AnyFunSuiteLike.run(AnyFunSuiteLike.scala:273)
	at org.scalatest.funsuite.AnyFunSuiteLike.run$(AnyFunSuiteLike.scala:272)
	at com.anonymized.cde.smmdevicetype.HudiWriteReadTest.org$scalatest$BeforeAndAfterAll$$super$run(HudiWriteReadTest.scala:15)
	at org.scalatest.BeforeAndAfterAll.liftedTree1$1(BeforeAndAfterAll.scala:213)
	at org.scalatest.BeforeAndAfterAll.run(BeforeAndAfterAll.scala:210)
	at org.scalatest.BeforeAndAfterAll.run$(BeforeAndAfterAll.scala:208)
	at com.anonymized.cde.smmdevicetype.HudiWriteReadTest.run(HudiWriteReadTest.scala:15)
	at org.scalatestplus.junit.JUnitRunner.run(JUnitRunner.scala:99)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:112)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:40)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:60)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:52)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$2.run(TestWorker.java:176)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:113)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:65)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
Caused by: java.lang.NoSuchMethodException: org.apache.hudi.metrics.cloudwatch.CloudWatchMetricsReporter.<init>(org.apache.hudi.config.metrics.HoodieMetricsConfig, com.codahale.metrics.MetricRegistry)
	at java.lang.Class.getConstructor0(Class.java:3110)
	at java.lang.Class.getConstructor(Class.java:1853)
	at org.apache.hudi.common.util.ReflectionUtils.loadClass(ReflectionUtils.java:73)
	... 106 more

```

--- a/hudi-common/src/main/java/org/apache/hudi/metrics/MetricsReporterFactory.java
+++ b/hudi-common/src/main/java/org/apache/hudi/metrics/MetricsReporterFactory.java
@@ -106,8 +106,9 @@ public class MetricsReporterFactory {
 
   /**
    * The CloudWatch reporter ships in the optional {@code hudi-aws} module and so is loaded reflectively.
-   * Not every engine bundle shades that module, in which case class loading fails without pointing at a
-   * remedy. Translate that into an actionable error.
+   * Reflection reports its two distinct failures - the module is absent, or the module is present but was
+   * built against a different Hudi version - as the same opaque {@link HoodieException}. Translate each into
+   * an error that names its own remedy, and leave anything else untouched.
    */
   private static MetricsReporter createCloudWatchReporter(HoodieMetricsConfig metricsConfig, MetricRegistry registry) {
     try {
@@ -121,6 +122,18 @@ public class MetricsReporterFactory {
                 + "hudi-aws-bundle jar matching your Hudi version to the classpath, or set %s to a "
                 + "different reporter type.",
             CLOUDWATCH_REPORTER_CLASS, HoodieMetricsConfig.METRICS_REPORTER_TYPE_VALUE.key()), e);
+      }
+      if (e.getCause() instanceof NoSuchMethodException) {
+        // The class resolved, so hudi-aws is present; only the constructor did not match. That means the
+        // jar providing it was built against a different Hudi version, which is a mismatch rather than
+        // something missing, and the two need different remedies.
+        throw new HoodieException(String.format(
+            "Cannot report metrics to CloudWatch: %s was found on the classpath but has no (%s, %s) "
+                + "constructor. The jar providing it was most likely built against a different Hudi "
+                + "version. Use a hudi-aws-bundle whose version matches the Hudi bundle in use, or set %s "
+                + "to a different reporter type.",
+            CLOUDWATCH_REPORTER_CLASS, HoodieMetricsConfig.class.getSimpleName(),
+            MetricRegistry.class.getSimpleName(), HoodieMetricsConfig.METRICS_REPORTER_TYPE_VALUE.key()), e);
       }
       throw e;
     }

--- a/hudi-common/src/main/java/org/apache/hudi/metrics/MetricsReporterFactory.java
+++ b/hudi-common/src/main/java/org/apache/hudi/metrics/MetricsReporterFactory.java
@@ -106,14 +106,33 @@ public class MetricsReporterFactory {
 
   /**
    * The CloudWatch reporter ships in the optional {@code hudi-aws} module and so is loaded reflectively.
-   * Reflection reports its two distinct failures - the module is absent, or the module is present but was
-   * built against a different Hudi version - as the same opaque {@link HoodieException}. Translate each into
-   * an error that names its own remedy, and leave anything else untouched.
+   * Reflection collapses several unrelated failures into the same opaque {@link HoodieException}. Three of
+   * them have distinct remedies - the module is absent, it was built against a Hudi that has since moved a
+   * class, or the classpath carries a stale duplicate - so translate those three, and leave every other
+   * failure untouched.
    */
   private static MetricsReporter createCloudWatchReporter(HoodieMetricsConfig metricsConfig, MetricRegistry registry) {
+    return createCloudWatchReporter(CLOUDWATCH_REPORTER_CLASS, metricsConfig, registry);
+  }
+
+  @VisibleForTesting
+  static MetricsReporter createCloudWatchReporter(String reporterClass, HoodieMetricsConfig metricsConfig,
+                                                 MetricRegistry registry) {
     try {
-      return (MetricsReporter) ReflectionUtils.loadClass(CLOUDWATCH_REPORTER_CLASS,
+      return (MetricsReporter) ReflectionUtils.loadClass(reporterClass,
           new Class[] {HoodieMetricsConfig.class, MetricRegistry.class}, metricsConfig, registry);
+    } catch (NoClassDefFoundError e) {
+      // Class#getConstructor resolves the parameter types of every public constructor, not just the one
+      // asked for, so a jar built against an older Hudi fails here on a type that has since moved - not
+      // with a missing-constructor error. NoClassDefFoundError is an Error, so ReflectionUtils never wraps
+      // it and it arrives here uncaught. Its message names the type that vanished, which is the strongest
+      // evidence of skew available.
+      throw new HoodieException(String.format(
+          "Cannot report metrics to CloudWatch: %s was found on the classpath but was built against a "
+              + "different Hudi version - resolving its constructors needs %s, which this Hudi no longer "
+              + "provides. Use a hudi-aws-bundle of the same version as the engine bundle, or set %s to a "
+              + "different reporter type.",
+          reporterClass, e.getMessage(), HoodieMetricsConfig.METRICS_REPORTER_TYPE_VALUE.key()), e);
     } catch (HoodieException e) {
       if (e.getCause() instanceof ClassNotFoundException) {
         throw new HoodieException(String.format(
@@ -121,19 +140,24 @@ public class MetricsReporterFactory {
                 + "optional hudi-aws module, which not every engine bundle includes. Add the "
                 + "hudi-aws-bundle jar matching your Hudi version to the classpath, or set %s to a "
                 + "different reporter type.",
-            CLOUDWATCH_REPORTER_CLASS, HoodieMetricsConfig.METRICS_REPORTER_TYPE_VALUE.key()), e);
+            reporterClass, HoodieMetricsConfig.METRICS_REPORTER_TYPE_VALUE.key()), e);
       }
       if (e.getCause() instanceof NoSuchMethodException) {
-        // The class resolved, so hudi-aws is present; only the constructor did not match. That means the
-        // jar providing it was built against a different Hudi version, which is a mismatch rather than
-        // something missing, and the two need different remedies.
+        // The class resolved and so did every constructor's parameter types, yet none matched. A jar built
+        // against an older Hudi fails earlier, in the NoClassDefFoundError branch above, so what reaches
+        // here is a classpath carrying a stale or duplicate copy of this class or of its parameter types.
+        // Fully qualified names, because the package move and the shaded-codahale relocation are exactly
+        // what distinguishes the declared constructor from the requested one - under simple names both read
+        // as (HoodieMetricsConfig, MetricRegistry) and the error looks wrong.
         throw new HoodieException(String.format(
-            "Cannot report metrics to CloudWatch: %s was found on the classpath but has no (%s, %s) "
-                + "constructor. The jar providing it was most likely built against a different Hudi "
-                + "version. Use a hudi-aws-bundle whose version matches the Hudi bundle in use, or set %s "
-                + "to a different reporter type.",
-            CLOUDWATCH_REPORTER_CLASS, HoodieMetricsConfig.class.getSimpleName(),
-            MetricRegistry.class.getSimpleName(), HoodieMetricsConfig.METRICS_REPORTER_TYPE_VALUE.key()), e);
+            "Cannot report metrics to CloudWatch: %s was found on the classpath but does not declare a "
+                + "(%s, %s) constructor. Some jar on the classpath is supplying a stale or duplicate copy "
+                + "of this class or of its parameter types. Check for more than one Hudi version on the "
+                + "classpath - a leftover hudi-common or hudi-client-common is the usual culprit - and "
+                + "align every Hudi artifact, including the engine bundle, to one version. Or set %s to a "
+                + "different reporter type.",
+            reporterClass, HoodieMetricsConfig.class.getName(),
+            MetricRegistry.class.getName(), HoodieMetricsConfig.METRICS_REPORTER_TYPE_VALUE.key()), e);
       }
       throw e;
     }

--- a/hudi-common/src/test/java/org/apache/hudi/metrics/TestMetricsReporterFactory.java
+++ b/hudi-common/src/test/java/org/apache/hudi/metrics/TestMetricsReporterFactory.java
@@ -37,9 +37,11 @@ import org.mockito.MockedStatic;
 import org.mockito.Mockito;
 import org.mockito.junit.jupiter.MockitoExtension;
 
+import java.lang.reflect.InvocationTargetException;
 import java.util.Properties;
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
 import static org.junit.jupiter.api.Assertions.assertSame;
 import static org.junit.jupiter.api.Assertions.assertThrows;
 import static org.junit.jupiter.api.Assertions.assertTrue;
@@ -137,11 +139,13 @@ class TestMetricsReporterFactory {
   }
 
   /**
-   * The other direction, and the branch most likely to regress: a failure that is not a missing class must
-   * pass through untouched, so an unrelated instantiation error is never rewritten into "add hudi-aws-bundle".
+   * A missing constructor means hudi-aws resolved but was built against a different Hudi version, which is
+   * a mismatch rather than something absent. Reflection reports it as the same opaque HoodieException as a
+   * missing class, and the bare "Unable to instantiate class" that resulted took a maintainer reading the
+   * buried NoSuchMethodException to explain (#12902).
    */
   @Test
-  void metricsReporterFactoryLeavesNonClassNotFoundFailuresUntouched() {
+  void metricsReporterFactoryExplainsAConstructorMismatch() {
     when(metricsConfig.getMetricsReporterType()).thenReturn(MetricsReporterType.CLOUDWATCH);
     try (MockedStatic<ReflectionUtils> mockedStatic = Mockito.mockStatic(ReflectionUtils.class)) {
       mockedStatic.when(() -> ReflectionUtils.loadClass(
@@ -151,8 +155,34 @@ class TestMetricsReporterFactory {
 
       HoodieException exception = assertThrows(HoodieException.class,
           () -> MetricsReporterFactory.createReporter(metricsConfig, registry));
+      String message = exception.getMessage();
+      assertTrue(message.contains("constructor"),
+          () -> "The failure should say the constructor did not match, but was: " + message);
+      assertTrue(message.contains("different Hudi version"),
+          () -> "The failure should name version skew as the likely cause, but was: " + message);
+      assertFalse(message.contains("was not found on the classpath"),
+          () -> "A class that resolved must not be reported as missing, but was: " + message);
+    }
+  }
+
+  /**
+   * The other direction, and the branch most likely to regress: a failure that is neither a missing class
+   * nor a missing constructor must pass through untouched, so an error raised by the reporter's own
+   * constructor is never rewritten into a classpath diagnosis.
+   */
+  @Test
+  void metricsReporterFactoryLeavesOtherFailuresUntouched() {
+    when(metricsConfig.getMetricsReporterType()).thenReturn(MetricsReporterType.CLOUDWATCH);
+    try (MockedStatic<ReflectionUtils> mockedStatic = Mockito.mockStatic(ReflectionUtils.class)) {
+      mockedStatic.when(() -> ReflectionUtils.loadClass(
+          eq(MetricsReporterFactory.CLOUDWATCH_REPORTER_CLASS), any(Class[].class), eq(metricsConfig), eq(registry)))
+          .thenThrow(new HoodieException("Unable to instantiate class " + MetricsReporterFactory.CLOUDWATCH_REPORTER_CLASS,
+              new InvocationTargetException(new IllegalStateException("no AWS region configured"))));
+
+      HoodieException exception = assertThrows(HoodieException.class,
+          () -> MetricsReporterFactory.createReporter(metricsConfig, registry));
       assertEquals("Unable to instantiate class " + MetricsReporterFactory.CLOUDWATCH_REPORTER_CLASS,
-          exception.getMessage(), "A non-ClassNotFound failure must not be rewritten");
+          exception.getMessage(), "A failure that is neither cause must not be rewritten");
     }
   }
 

--- a/hudi-common/src/test/java/org/apache/hudi/metrics/TestMetricsReporterFactory.java
+++ b/hudi-common/src/test/java/org/apache/hudi/metrics/TestMetricsReporterFactory.java
@@ -182,7 +182,8 @@ class TestMetricsReporterFactory {
       HoodieException exception = assertThrows(HoodieException.class,
           () -> MetricsReporterFactory.createReporter(metricsConfig, registry));
       assertEquals("Unable to instantiate class " + MetricsReporterFactory.CLOUDWATCH_REPORTER_CLASS,
-          exception.getMessage(), "A failure that is neither cause must not be rewritten");
+          exception.getMessage(),
+          "A failure that is neither a missing class nor a missing constructor must not be rewritten");
     }
   }
 

--- a/hudi-common/src/test/java/org/apache/hudi/metrics/TestMetricsReporterFactory.java
+++ b/hudi-common/src/test/java/org/apache/hudi/metrics/TestMetricsReporterFactory.java
@@ -41,7 +41,6 @@ import java.lang.reflect.InvocationTargetException;
 import java.util.Properties;
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
-import static org.junit.jupiter.api.Assertions.assertFalse;
 import static org.junit.jupiter.api.Assertions.assertSame;
 import static org.junit.jupiter.api.Assertions.assertThrows;
 import static org.junit.jupiter.api.Assertions.assertTrue;
@@ -116,6 +115,10 @@ class TestMetricsReporterFactory {
         () -> "The failure should name the bundle that provides the reporter, but was: " + message);
     assertTrue(message.contains(HoodieMetricsConfig.METRICS_REPORTER_TYPE_VALUE.key()),
         () -> "The failure should name the config to change, but was: " + message);
+    // Every string above also appears in the constructor-mismatch message, so without this the two branches
+    // could be merged or reordered and both tests would still pass.
+    assertTrue(message.contains("was not found on the classpath"),
+        () -> "A missing class must not be reported as a constructor mismatch, but was: " + message);
   }
 
   /**
@@ -135,33 +138,85 @@ class TestMetricsReporterFactory {
           () -> MetricsReporterFactory.createReporter(metricsConfig, registry));
       assertTrue(exception.getMessage().contains("hudi-aws-bundle"),
           () -> "Expected the remedy to be named, but was: " + exception.getMessage());
+      assertTrue(exception.getMessage().contains("was not found on the classpath"),
+          () -> "Expected this branch's own phrase, not one shared with the mismatch branch, but was: "
+              + exception.getMessage());
     }
   }
 
   /**
-   * A missing constructor means hudi-aws resolved but was built against a different Hudi version, which is
-   * a mismatch rather than something absent. Reflection reports it as the same opaque HoodieException as a
-   * missing class, and the bare "Unable to instantiate class" that resulted took a maintainer reading the
-   * buried NoSuchMethodException to explain (#12902).
+   * A jar built against an older Hudi does not reach this branch - resolving its constructors fails first
+   * with {@link NoClassDefFoundError}, covered below. What reaches here is a classpath carrying a stale or
+   * duplicate copy, so that is the remedy the message has to give.
    */
   @Test
   void metricsReporterFactoryExplainsAConstructorMismatch() {
-    when(metricsConfig.getMetricsReporterType()).thenReturn(MetricsReporterType.CLOUDWATCH);
-    try (MockedStatic<ReflectionUtils> mockedStatic = Mockito.mockStatic(ReflectionUtils.class)) {
-      mockedStatic.when(() -> ReflectionUtils.loadClass(
-          eq(MetricsReporterFactory.CLOUDWATCH_REPORTER_CLASS), any(Class[].class), eq(metricsConfig), eq(registry)))
-          .thenThrow(new HoodieException("Unable to instantiate class " + MetricsReporterFactory.CLOUDWATCH_REPORTER_CLASS,
-              new NoSuchMethodException("<init>")));
+    HoodieException exception = captureCloudWatchFailure(
+        new HoodieException("Unable to instantiate class " + MetricsReporterFactory.CLOUDWATCH_REPORTER_CLASS,
+            new NoSuchMethodException("<init>")));
 
-      HoodieException exception = assertThrows(HoodieException.class,
-          () -> MetricsReporterFactory.createReporter(metricsConfig, registry));
-      String message = exception.getMessage();
-      assertTrue(message.contains("constructor"),
-          () -> "The failure should say the constructor did not match, but was: " + message);
-      assertTrue(message.contains("different Hudi version"),
-          () -> "The failure should name version skew as the likely cause, but was: " + message);
-      assertFalse(message.contains("was not found on the classpath"),
-          () -> "A class that resolved must not be reported as missing, but was: " + message);
+    String message = exception.getMessage();
+    assertTrue(message.contains("constructor"),
+        () -> "The failure should say the constructor did not match, but was: " + message);
+    assertTrue(message.contains("stale or duplicate copy"),
+        () -> "The failure should name a stale duplicate as the cause, but was: " + message);
+    assertTrue(message.contains(MetricsReporterFactory.CLOUDWATCH_REPORTER_CLASS),
+        () -> "The failure should name the reporter class, but was: " + message);
+    assertTrue(message.contains(HoodieMetricsConfig.METRICS_REPORTER_TYPE_VALUE.key()),
+        () -> "The failure should name the config to change, but was: " + message);
+    assertTrue(message.contains(HoodieMetricsConfig.class.getName())
+            && message.contains(MetricRegistry.class.getName()),
+        () -> "Fully qualified parameter types are what distinguish the requested constructor from the "
+            + "declared one, but was: " + message);
+    // Positive rather than an assertFalse on the other branch's prose: a vacuous assertFalse never fails.
+    assertTrue(message.contains("was found on the classpath but"),
+        () -> "A class that resolved must not be reported as missing, but was: " + message);
+    assertEquals(NoSuchMethodException.class, exception.getCause().getCause().getClass(),
+        "The original NoSuchMethodException must stay in the chain - it is the evidence #12902 needed");
+  }
+
+  /**
+   * The clean version-skew case, and the one the mismatch message used to claim. {@code getConstructor}
+   * resolves the parameter types of every public constructor, so a jar built against an older Hudi dies on a
+   * type that has since moved. That is an {@link Error}, so {@code ReflectionUtils} never wraps it and it
+   * reaches the factory uncaught.
+   */
+  @Test
+  void metricsReporterFactoryExplainsAVanishedParameterType() {
+    HoodieException exception = captureCloudWatchFailure(
+        new NoClassDefFoundError("org/apache/hudi/config/metrics/HoodieMetricsConfig"));
+
+    String message = exception.getMessage();
+    assertTrue(message.contains("built against a different Hudi version"),
+        () -> "The failure should name version skew, but was: " + message);
+    assertTrue(message.contains("org/apache/hudi/config/metrics/HoodieMetricsConfig"),
+        () -> "The failure should name the type that vanished, but was: " + message);
+    assertEquals(NoClassDefFoundError.class, exception.getCause().getClass(),
+        "The Error must stay in the chain - its message is the evidence of skew");
+  }
+
+  /**
+   * The gap every mocked test above shares: none of them proves that real reflection produces the shapes the
+   * production code branches on. This drives the translation through the real {@code ReflectionUtils} with a
+   * fixture whose only public constructor does not match, and asserts on the actual throwable.
+   */
+  @Test
+  void metricsReporterFactoryTranslatesARealReflectionFailure() {
+    HoodieException exception = assertThrows(HoodieException.class,
+        () -> MetricsReporterFactory.createCloudWatchReporter(
+            MismatchedReporter.class.getName(), metricsConfig, registry));
+
+    String message = exception.getMessage();
+    assertTrue(message.contains("stale or duplicate copy"),
+        () -> "A real non-matching constructor should reach the mismatch branch, but was: " + message);
+    assertEquals(NoSuchMethodException.class, exception.getCause().getCause().getClass(),
+        "and the real NoSuchMethodException should be chained");
+  }
+
+  /** Public, with a single constructor that deliberately does not match (HoodieMetricsConfig, MetricRegistry). */
+  public static class MismatchedReporter {
+    public MismatchedReporter(String somethingElse) {
+      // never called; exists so getConstructor has a public constructor to reject
     }
   }
 
@@ -206,6 +261,17 @@ class TestMetricsReporterFactory {
     when(metricsConfig.getMetricReporterClassName()).thenReturn(IllegalTestMetricsReporter.class.getName());
     when(metricsConfig.getProps()).thenReturn(new TypedProperties());
     assertThrows(HoodieException.class, () -> MetricsReporterFactory.createReporter(metricsConfig, registry));
+  }
+
+  private HoodieException captureCloudWatchFailure(Throwable reflectionFailure) {
+    when(metricsConfig.getMetricsReporterType()).thenReturn(MetricsReporterType.CLOUDWATCH);
+    try (MockedStatic<ReflectionUtils> mockedStatic = Mockito.mockStatic(ReflectionUtils.class)) {
+      mockedStatic.when(() -> ReflectionUtils.loadClass(
+          eq(MetricsReporterFactory.CLOUDWATCH_REPORTER_CLASS), any(Class[].class), eq(metricsConfig), eq(registry)))
+          .thenThrow(reflectionFailure);
+      return assertThrows(HoodieException.class,
+          () -> MetricsReporterFactory.createReporter(metricsConfig, registry));
+    }
   }
 
   public static class DummyMetricsReporter extends CustomizableMetricsReporter {

